### DISCUSSION TITLE
Minimum PHP version for Yii 2.2

### DISCUSSION THREAD

**Discussion by rhertogh** [yiisoft/yii2#19878] [Polls]
Based on the discussion in https://github.com/yiisoft/yii2/discussions/19831 and https://github.com/yiisoft/yii2/discussions/19875 I would like propose a poll.
Please read the arguments (pros/cons) below and add your own.

Current (2023-06-27) [usage statistics](https://packagist.org/packages/yiisoft/yii2/php-stats#all):

![image](https://github.com/yiisoft/yii2/assets/1292337/e7f20be7-7136-473b-8cc6-2af4552ee171)

**From the [Reusing Yii3 packages](https://github.com/yiisoft/yii2/discussions/19875) discussion:**

@samdark:
> The hard decision here is to align PHP versions so that will require Yii 2.2 to be PHP 8.0+ at least.

**From the [Yii 2.2](https://github.com/yiisoft/yii2/discussions/19831) discussion:**

@tomaszkane:
> IMHO as @schmunk42 says: https://packagist.org/packages/yiisoft/yii2/php-stats PHP 5.4 is 0.0 - 0.1% installations - so (almost) no one needs support this version.
> For me resonable minimum is last "7" branch: 7.4.

@My6UoT9:
> I would like to see Yii2 being in sync with supported PHP versions, so that would be 8,1+
> edit: actually php 8.0 since that is suported till end of year. https://www.php.net/supported-versions.php

@schmunk42:
> I'd advocate for 8.1 - I upgraded a lot of projects in the past months and while upgrading to 8.0 is rather easy, 8.1 and 8.2 raises a lot more of issues.

@mtangoo:
> In our company we are running on 8.1 with no issues. 8.2 is rather a problem.

@rhertogh :
> I would also advocate for PHP 8.1. It's already supported in Yii 2.0, so anyone who need a lower PHP version can stay on that version and step by step transition to PHP 8.1. From there they could make the switch to Yii 2.2 if they want.

@samdark:
> hhvm and PHP 5 are safe to drop support for.
pre-7.4 versions are 19% together and that is signficiant. I wonder if these are using older Yii2 versions as well but we don't have the stats.
> 7.4 is the most popular version despite it being EOL-ed.

@terabytesoftw:
> This has turned into a snowball, the more versions of PHP are released, the more difficult it will be to have compatibility, so i think it should be done, with a minimum version of at least 8.0 or 8.1, whoever uses previous versions will stay at 2.0.

@samdark:
> Some reasons on why people don't switch to PHP 8: https://twitter.com/sam_dark/status/1656549413282611200

@rhertogh:
> I think the current users can be grouped in 3 groups:
>
> 1. Users that want to move forward.
>    - Willing to upgrade to Yii 2.2
>    - Use a recent PHP version (>= 8.1)
> 2. Users that are happy to stay where they are.
>    - Are OK with staying on Yii 2.0
>    - Probably are using an older PHP version (e.g. 7.4)
> 3. Users that would want to use Yii 2.2 but can't.
>    - Would like to upgrade to Yii 2.2
>    - Stuck on an older PHP version (e.g. 7.4)
> 
> So my argument is that even a lot of the current Yii 2 installs are on PHP < 8.1 they are most likely in group 2 (happy to stay). Therefore the question should be: Should group 1 be hold back by group 3?
> Since I don't think group 3 makes up a significant percentage I would argue that Yii 2.2 should focus on group 1 and use PHP 8.1 as the minimum version.

p.s. I tried to summarize the arguments as good as possible. If anyone feel they are misquoted or want their argument updated, please let me know.

**Comment by tomaszkane** [comment]
Update: IMHO it should be like in Yii3, latest supported PHP version.
When 2.2 will be ready it will be 8.1 I believe.
It's new Yii version, not simple update.

**Comment by schmunk42** [comment]
If an intermediate step with PHP 7.4 as a minimum requirement helps (see https://github.com/yiisoft/yii2/discussions/19831#discussioncomment-5869208) I'd be fine with it, just sayin'.
But you'd need a Yii 2.3 version for PHP 8+ then I think.

**Reply by mtangoo** [reply]
I think we can keep 2.0.x as <=7.4 bug-fixing/sec update only while >=2.1 will be 8.1+ only. So one will have to choose.
8.0 is not an issue for me as I got a big system from 7.4 to 8.0 to 8.1 with almost no remembable change. So  we can count 8.0 as 8.1 vote (without sounding dictators ;))

**Comment by terabytesoftw** [comment]
I'm working on a draft of php `8.0`. and it's really not easy, getting everything working will take some time, so it's best to do this effort on a newer version of PHP.

And if it were better php `8.1` is better, `PDO` changes from `8.0` to `8.1`, in `8.0` all `int` are `string` in `8.1` they are `int` :)

**Reply by rhertogh** [reply]
How do you mean that exactly, isn't Yii 2.0 already compatible with PHP 8.1?

**Reply by terabytesoftw** [reply]
Yes it is compatible, but some small adjustments must be made.

**Comment by samdark** [comment]
Here's a fresh update on PHP version stats: https://stitcher.io/blog/php-version-stats-july-2023 Seems 8.1 is growing fast.

**Reply by mtangoo** [reply]
- 7.4 ie EOL and in the fall
- 8.0 + 8.1 accounts for more than 50%
- 8.x are rising
I would say IMHO, that 8.0/8.1 are best bet on minimum PHP version reading this chart

**Reply by lourdas** [reply]
I don't get why 8.0 never reached above 24%... Yes, 8.1 is indeed growing fast.

FWIW, I voted for 8.0, since there are servers out there that must serve a few applications at the same time, so compatibility with 8.1 might be something that just doesn't work.

**Reply by mtangoo** [reply]
If it is shared servers, then you can set individual subdomains/domains PHP versions in CPanel.
If it is VPS, you can have multiple instances of apache running on different ports, if that is an option

**Comment by xicond** [comment]
Do we need add php ver limitation in v2.0, let say <8.0

**Reply by bizley** [reply]
No.

**Comment by terabytesoftw** [comment]
done.
